app/src/org/commcare/CommCareHeartbeatManager.java

```diff
@@ -0,0 +1,309 @@
+        try {
+            if (latestVersionInfo.has("value")) {
+                String versionValue = latestVersionInfo.getString("value");
+                String forceByDate = null;
+                if (latestVersionInfo.has("force_by_date")) {
+                    forceByDate = latestVersionInfo.getString("force_by_date");
+                }
+                UpdateToPrompt updateToPrompt = new UpdateToPrompt(versionValue, forceByDate, isForApk);
+                updateToPrompt.registerWithSystem();
+            }
+        } catch (JSONException e) {
+            Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                    "Encountered malformed json while trying to parse server response into an " +
+                            "UpdateToPrompt object : " + e.getMessage());
+        }
+    }
+
+    /**
+     * @return - If the user was prompted to update
+     */
+    public static boolean promptForUpdateIfNeeded(Activity context) {
+        UpdateToPrompt cczUpdate = getCurrentUpdateToPrompt(false);
+        UpdateToPrompt apkUpdate = getCurrentUpdateToPrompt(true);
+        if (cczUpdate != null || apkUpdate != null) {
+            Intent i = new Intent(context, PromptUpdateActivity.class);
+            context.startActivity(i);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @return an UpdateToPrompt that has been stored in SharedPreferences and is still relevant
+     * (i.e. the user hasn't updated to or past this version since we stored it)
+     */
+    public static UpdateToPrompt getCurrentUpdateToPrompt(boolean forApkUpdate) {
+        CommCareApp currentApp = CommCareApplication.instance().getCurrentApp();
+        if (currentApp != null) {
+            String prefsKey = forApkUpdate ?
+                    UpdateToPrompt.KEY_APK_UPDATE_TO_PROMPT : UpdateToPrompt.KEY_CCZ_UPDATE_TO_PROMPT;
+            String serializedUpdate = currentApp.getAppPreferences().getString(prefsKey, "");
+            if (!"".equals(serializedUpdate)) {
+                try {
+                    byte[] updateBytes = Base64.decode(serializedUpdate, Base64.DEFAULT);
+                    DataInputStream stream = new DataInputStream(new ByteArrayInputStream(updateBytes));
+                    UpdateToPrompt update = (UpdateToPrompt)
+                            ExtUtil.read(stream, UpdateToPrompt.class, ExtUtil.defaultPrototypes());
+                    if (update.isNewerThanCurrentVersion(currentApp)) {
+                        return update;
+                    } else {
+                        // The update we had stored is no longer relevant, so wipe it and return nothing
+                        wipeStoredUpdate(forApkUpdate);
+                        return null;
+                    }
+                } catch (Exception e) {
+                    // Something went wrong, so clear out whatever is there
+                    Logger.log(AndroidLogger.TYPE_ERROR_WORKFLOW,
```

Comment:
Ah, I think I was unclear in this comment, the issue was't around why deserialization failed, it was that there was more code wrapped in the try/catch than just the read/deserialize, so in theory there could be other bugs in the code that instead of being properly thrown would get obfuscated as deserialization errors.